SwiftGit2/Repository.swift

```diff
@@ -537,4 +537,142 @@ final public class Repository {
 		let iterator = CommitIterator(repo: self, root: branch.oid.oid)
 		return iterator
 	}
+
+	// MARK: - Diffs
+
+	public func diff(for commit: Commit) -> Result<[DiffDelta], NSError> {
+		/// Get the Base Tree
+		var unsafeBaseCommit: OpaquePointer? = nil
+		let unsafeBaseOid = UnsafeMutablePointer<git_oid>.allocate(capacity: 1)
+		git_oid_fromstr(unsafeBaseOid, commit.oid.description)
+		let lookupBaseGitResult = git_commit_lookup(&unsafeBaseCommit, self.pointer, unsafeBaseOid)
+		guard lookupBaseGitResult == GIT_OK.rawValue, let unwrapBaseCommit = unsafeBaseCommit else {
+				return Result.failure(NSError(gitError: lookupBaseGitResult, pointOfFailure: "git_commit_lookup"))
+		}
+		git_commit_free(unsafeBaseCommit)
+
+		var unsafeBaseTree: OpaquePointer? = nil
+		let baseTreeResult = git_commit_tree(&unsafeBaseTree, unwrapBaseCommit)
+		guard baseTreeResult == GIT_OK.rawValue, let unwrapBaseTree = unsafeBaseTree else {
+			return Result.failure(NSError(gitError: baseTreeResult, pointOfFailure: "git_commit_tree"))
+		}
+		git_tree_free(unsafeBaseTree)
+
+		if commit.parents.isEmpty {
+			// Initial commit in a repository
+			var unsafeDiff: OpaquePointer? = nil
+			let diffResult = git_diff_tree_to_tree(&unsafeDiff, self.pointer, nil, unwrapBaseTree, nil)
+			guard diffResult == GIT_OK.rawValue, let unwrapDiffResult = unsafeDiff else {
+				return Result.failure(NSError(gitError: diffResult, pointOfFailure: "git_diff_tree_to_tree"))
+			}
+
+			return self.processDiffDeltas(unwrapDiffResult)
+		} else {
+			// Possible Merge Commit, merge diffs of base with each parent
+			var mergeDiff: OpaquePointer? = nil
+			for parent in commit.parents {
+				var unsafeParentCommit: OpaquePointer? = nil
+				let unsafeParentOid = UnsafeMutablePointer<git_oid>.allocate(capacity: 1)
+				git_oid_fromstr(unsafeParentOid, parent.oid.description)
+				let lookupParentGitResult = git_commit_lookup(&unsafeParentCommit, self.pointer, unsafeParentOid)
+				guard lookupParentGitResult == GIT_OK.rawValue, let unwrapParentCommit = unsafeParentCommit else {
+					return Result.failure(NSError(gitError: lookupParentGitResult, pointOfFailure: "git_commit_lookup"))
+				}
+				git_commit_free(unsafeParentCommit)
+
+				var unsafeParentTree: OpaquePointer? = nil
+				let parentTreeResult = git_commit_tree(&unsafeParentTree, unwrapParentCommit)
+				guard parentTreeResult == GIT_OK.rawValue, let unwrapParentTree = unsafeParentTree else {
+					return Result.failure(NSError(gitError: parentTreeResult, pointOfFailure: "git_commit_tree"))
+				}
+				git_tree_free(unsafeParentTree)
+
+				var unsafeDiff: OpaquePointer? = nil
+				let diffResult = git_diff_tree_to_tree(&unsafeDiff, self.pointer, unwrapParentTree, unwrapBaseTree, nil)
+				guard diffResult == GIT_OK.rawValue, let unwrapDiffResult = unsafeDiff else {
+					return Result.failure(NSError(gitError: diffResult, pointOfFailure: "git_diff_tree_to_tree"))
```

Comment:
`git_diff_tree_to_tree`, `git_commit_lookup`, and `git_commit_tree` are all used multiple times, but they're quite a bit of code. It seems like we could create helper functions for these? That should really improve readability here.